File: services/api/app/services/reader_orchestration/semantic_outline.py
"""Pure, fail-closed semantic-outline validator."""

from __future__ import annotations

import unicodedata

from dataclasses import dataclass
from typing import Mapping

from app.schemas.reader_orchestration import (
    ReaderSemanticOutlineDiagnostics,
    ReaderSemanticOutlineDrop,
    ReaderSemanticOutlineNode,
    ReaderSemanticOutlineStatus,
)


SemanticOutlineStatus = ReaderSemanticOutlineStatus
_MAX_DEPTH = 3
_TITLE_MAX_CODE_POINTS = 80
_ES_TRIM_EXPLICIT_CODE_POINTS = frozenset({0x0009, 0x000A, 0x000B, 0x000C, 0x000D, 0x0020, 0x00A0, 0x2028, 0x2029, 0xFEFF})


def _mapping(value: object) -> Mapping[str, object]:
    return value if isinstance(value, Mapping) else {}


def _optional_string(value: object) -> str | None:
    return value if isinstance(value, str) and value else None


def _es_trim(value: str) -> str:
    def is_trim_character(character: str) -> bool:
        return ord(character) in _ES_TRIM_EXPLICIT_CODE_POINTS or unicodedata.category(character) == "Zs"

    start, end = 0, len(value)
    while start < end and is_trim_character(value[start]):
        start += 1
    while end > start and is_trim_character(value[end - 1]):
        end -= 1
    return value[start:end]


@dataclass(frozen=True)
class SemanticOutlineSourceIdentity:
    base_id: str
    generation: int

    @classmethod
    def from_mapping(cls, value: object) -> SemanticOutlineSourceIdentity:
        mapping = _mapping(value)
        generation = mapping.get("generation")
        return cls(
            base_id=_optional_string(mapping.get("base_id")) or "",
            generation=generation if isinstance(generation, int) and not isinstance(generation, bool) else 0,
        )


@dataclass(frozen=True)
class SemanticOutlineUnit:
    unit_id: str
    order_index: int


@dataclass(frozen=True)
class SemanticOutlineAnchor:
    anchor_segment_id: str
    unit_id: str


@dataclass(frozen=True)
class SemanticOutlineValidationContext:
    source_identity: SemanticOutlineSourceIdentity
    units: tuple[SemanticOutlineUnit, ...]
    anchors: tuple[SemanticOutlineAnchor, ...]

    @classmethod
    def from_mapping(cls, value: object) -> SemanticOutlineValidationContext:
        mapping = _mapping(value)
        raw_units = mapping.get("units")
        units = tuple(
            SemanticOutlineUnit(unit_id=unit_id, order_index=order_index)
            for item in raw_units if isinstance(raw_units, list)
            for raw in (_mapping(item),)
            for unit_id in (_optional_string(raw.get("unit_id")),)
            for order_index in (raw.get("order_index"),)
            if unit_id is not None and isinstance(order_index, int) and not isinstance(order_index, bool) and order_index >= 1
        )
        raw_anchors = mapping.get("anchors")
        anchors = tuple(
            SemanticOutlineAnchor(anchor_segment_id=anchor_id, unit_id=unit_id)
            for item in raw_anchors if isinstance(raw_anchors, list)
            for raw in (_mapping(item),)
            for anchor_id in (_optional_string(raw.get("anchor_segment_id")),)
            for unit_id in (_optional_string(raw.get("unit_id")),)
            if anchor_id is not None and unit_id is not None
        )
        return cls(SemanticOutlineSourceIdentity.from_mapping(mapping.get("source_identity")), units, anchors)


@dataclass(frozen=True)
class RawSemanticOutlineNode:
    node_id: str | None
    parent_node_id: str | None
    depth: object
    title: object
    start_unit_id: str | None
    end_unit_id: str | None
    start_anchor_segment_id: str | None
    end_anchor_segment_id: str | None

    @classmethod
    def from_mapping(cls, value: object) -> RawSemanticOutlineNode:
        raw = _mapping(value)
        return cls(
            _optional_string(raw.get("node_id")), _optional_string(raw.get("parent_node_id")), raw.get("depth"), raw.get("title"),
            _optional_string(raw.get("start_unit_id")), _optional_string(raw.get("end_unit_id")),
            _optional_string(raw.get("start_anchor_segment_id")), _optional_string(raw.get("end_anchor_segment_id")),
        )


@dataclass(frozen=True)
class SemanticOutlineValidationInput:
    field_present: bool
    requested: bool
    in_flight: bool
    worker_failure: bool
    projection_source_identity: SemanticOutlineSourceIdentity
    attempted_nodes: tuple[RawSemanticOutlineNode, ...]

    @classmethod
    def from_mapping(cls, value: object) -> SemanticOutlineValidationInput:
        mapping = _mapping(value)
        raw_nodes = mapping.get("attempted_nodes")
        nodes = tuple(RawSemanticOutlineNode.from_mapping(item) for item in raw_nodes if isinstance(raw_nodes, list))
        return cls(mapping.get("field_present") is True, mapping.get("requested") is True, mapping.get("in_flight") is True, mapping.get("worker_failure") is True, SemanticOutlineSourceIdentity.from_mapping(mapping.get("projection_source_identity")), nodes)


SemanticOutlineDrop = ReaderSemanticOutlineDrop
SemanticOutlineDiagnostics = ReaderSemanticOutlineDiagnostics
ValidatedSemanticOutlineNode = ReaderSemanticOutlineNode


@dataclass(frozen=True)
class SemanticOutlineValidationResult:
    status: SemanticOutlineStatus
    nodes: tuple[ValidatedSemanticOutlineNode, ...]
    diagnostics: SemanticOutlineDiagnostics


def validate_semantic_outline_projection(
    context: SemanticOutlineValidationContext,
    validation_input: SemanticOutlineValidationInput,
) -> SemanticOutlineValidationResult:
    """Validate candidate preorder with source fence and fail-closed tree closure."""

    def make_diagnostics(
        drops: list[SemanticOutlineDrop], skipped_node_count: int
    ) -> SemanticOutlineDiagnostics:
        return SemanticOutlineDiagnostics(
            drops=drops,
            skipped_node_count=skipped_node_count,
        )

    def terminal(
        status: SemanticOutlineStatus, reason: str | None = None
    ) -> SemanticOutlineValidationResult:
        drops = [] if reason is None else [
            SemanticOutlineDrop(node_id=None, reason_code=reason)
        ]
        return SemanticOutlineValidationResult(
            status,
            (),
            make_diagnostics(drops, 0),
        )

    attempted = validation_input.attempted_nodes
    if validation_input.worker_failure:
        return terminal("failed", "worker_failure")
    if not validation_input.field_present or (
        not validation_input.requested and not attempted
    ):
        return terminal("unavailable")
    if validation_input.in_flight and not attempted:
        return terminal("pending")
    if validation_input.projection_source_identity != context.source_identity:
        return terminal("stale", "source_mismatch")
    if not attempted:
        return terminal("failed", "empty_attempt")

    unit_order = {unit.unit_id: unit.order_index for unit in context.units}
    anchor_units = {anchor.anchor_segment_id: anchor.unit_id for anchor in context.anchors}
    accepted: list[ValidatedSemanticOutlineNode] = []
    accepted_by_id: dict[str, ValidatedSemanticOutlineNode] = {}
    dropped_ids: set[str] = set()
    seen_ids: set[str] = set()
    drops: list[SemanticOutlineDrop] = []

    def drop(
        raw: RawSemanticOutlineNode, reason: str, *, mark_dropped: bool = True
    ) -> None:
        drops.append(SemanticOutlineDrop(node_id=raw.node_id, reason_code=reason))
        if mark_dropped and raw.node_id is not None:
            dropped_ids.add(raw.node_id)

    for raw in attempted:
        if raw.node_id is None:
            drop(raw, "missing_node_id")
            continue
        if raw.node_id in seen_ids:
            # A duplicate raw candidate must not poison an already accepted parent.
            drop(raw, "duplicate_node_id", mark_dropped=False)
            continue
        seen_ids.add(raw.node_id)
        parent = raw.parent_node_id
        parent_node = accepted_by_id.get(parent) if parent is not None else None
        if parent is not None:
            if parent in dropped_ids:
                drop(raw, "parent_dropped")
                continue
            if parent_node is None:
                drop(raw, "invalid_parent")
                continue
            if raw.depth != parent_node.depth + 1:
                drop(raw, "depth_parent_mismatch")
                continue
        if (
            not isinstance(raw.depth, int)
            or isinstance(raw.depth, bool)
            or not 1 <= raw.depth <= _MAX_DEPTH
        ):
            drop(raw, "depth_out_of_range")
            continue
        if parent is None and raw.depth != 1:
            drop(raw, "invalid_root_depth")
            continue
        if not isinstance(raw.title, str):
            drop(raw, "empty_title")
            continue
        title = _es_trim(raw.title)
        if not title:
            drop(raw, "empty_title")
            continue
        if len(title) > _TITLE_MAX_CODE_POINTS:
            drop(raw, "title_too_long")
            continue
        if raw.start_unit_id not in unit_order or raw.end_unit_id not in unit_order:
            drop(raw, "missing_unit")
            continue
        start_order = unit_order[raw.start_unit_id]
        end_order = unit_order[raw.end_unit_id]
        if start_order > end_order:
            drop(raw, "inverted_range")
            continue
        if parent_node is not None:
            parent_start = unit_order[parent_node.start_unit_id]
            parent_end = unit_order[parent_node.end_unit_id]
            if start_order < parent_start or end_order > parent_end:
                drop(raw, "range_not_nested")
                continue
        if any(
            existing.parent_node_id == parent
            and not (
                end_order < unit_order[existing.start_unit_id]
                or start_order > unit_order[existing.end_unit_id]
            )
            for existing in accepted
        ):
            drop(raw, "range_overlap")
            continue
        start_anchor = raw.start_anchor_segment_id
        end_anchor = raw.end_anchor_segment_id
        invalid_anchor = False
        if (
            start_anchor is not None
            and anchor_units.get(start_anchor) != raw.start_unit_id
        ):
            start_anchor = None
            invalid_anchor = True
        if end_anchor is not None and anchor_units.get(end_anchor) != raw.end_unit_id:
            end_anchor = None
            invalid_anchor = True
        if invalid_anchor:
            drops.append(
                SemanticOutlineDrop(node_id=raw.node_id, reason_code="invalid_anchor")
            )
        node = ValidatedSemanticOutlineNode(
            node_id=raw.node_id,
            parent_node_id=parent,
            depth=raw.depth,
            title=title,
            start_unit_id=raw.start_unit_id,
            end_unit_id=raw.end_unit_id,
            start_anchor_segment_id=start_anchor,
            end_anchor_segment_id=end_anchor,
            order_index=len(accepted) + 1,
        )
        accepted.append(node)
        accepted_by_id[node.node_id] = node

    valid_count = len(accepted)
    attempted_count = len(attempted)
    status: SemanticOutlineStatus
    if attempted_count == 0 or valid_count == 0:
        status = "failed"
    elif valid_count == attempted_count:
        status = "ready"
    else:
        status = "partial"
    return SemanticOutlineValidationResult(
        status,
        tuple(accepted),
        make_diagnostics(drops, attempted_count - valid_count),
    )

File: apps/web/src/lib/reader-plate/projection/semantic-outline.ts
/** Backend-owned semantic-outline wire consumer and pure validator. */

export type SemanticOutlineStatus =
  | "unavailable"
  | "pending"
  | "partial"
  | "ready"
  | "failed"
  | "stale";

export type ReaderSemanticOutlineSourceIdentityDto = {
  base_id: string;
  generation: number;
};
export type ReaderSemanticOutlinePublicationDto = {
  outline_revision: string;
  layer_id?: string | null;
  published_at?: string | null;
};
export type ReaderSemanticOutlineProvenanceDto = {
  kind: "llm" | "hybrid" | "deterministic";
  builder?: string | null;
  model?: string | null;
};
export type ReaderSemanticOutlineNodeDto = {
  node_id: string;
  parent_node_id: string | null;
  depth: number;
  title: string;
  start_unit_id: string;
  end_unit_id: string;
  start_anchor_segment_id: string | null;
  end_anchor_segment_id: string | null;
  order_index: number;
};
export type ReaderSemanticOutlineDropDto = {
  node_id: string | null;
  reason_code: string;
};
export type ReaderSemanticOutlineDiagnosticsDto = {
  drops: ReaderSemanticOutlineDropDto[];
  skipped_node_count: number;
};
export type ReaderSemanticOutlineProjectionDto = {
  schema_kind: "reader_semantic_outline";
  schema_version: 1;
  status: SemanticOutlineStatus;
  source_identity: ReaderSemanticOutlineSourceIdentityDto;
  publication: ReaderSemanticOutlinePublicationDto;
  provenance: ReaderSemanticOutlineProvenanceDto;
  nodes: ReaderSemanticOutlineNodeDto[];
  diagnostics: ReaderSemanticOutlineDiagnosticsDto;
};

export type SemanticOutlineSourceIdentity = ReaderSemanticOutlineSourceIdentityDto;
export type SemanticOutlineValidationContext = {
  source_identity: SemanticOutlineSourceIdentity;
  units: Array<{ unit_id: string; order_index: number }>;
  anchors: Array<{ anchor_segment_id: string; unit_id: string }>;
};
type RawSemanticOutlineNode = {
  node_id: string | null;
  parent_node_id: string | null;
  depth: unknown;
  title: unknown;
  start_unit_id: string | null;
  end_unit_id: string | null;
  start_anchor_segment_id: string | null;
  end_anchor_segment_id: string | null;
};
export type SemanticOutlineValidationInput = {
  field_present: boolean;
  requested: boolean;
  in_flight: boolean;
  worker_failure: boolean;
  projection_source_identity: SemanticOutlineSourceIdentity;
  attempted_nodes: RawSemanticOutlineNode[];
};
export type ValidatedSemanticOutlineNode = ReaderSemanticOutlineNodeDto;
export type SemanticOutlineDrop = ReaderSemanticOutlineDropDto;
export type SemanticOutlineValidationResult = {
  status: SemanticOutlineStatus;
  nodes: ValidatedSemanticOutlineNode[];
  diagnostics: ReaderSemanticOutlineDiagnosticsDto;
};

const MAX_DEPTH = 3;
const TITLE_MAX_CODE_POINTS = 80;

function mapping(value: unknown): Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : {};
}

function optionalString(value: unknown): string | null {
  return typeof value === "string" && value.length > 0 ? value : null;
}

function sourceIdentity(value: unknown): SemanticOutlineSourceIdentity {
  const raw = mapping(value);
  return {
    base_id: optionalString(raw.base_id) ?? "",
    generation:
      typeof raw.generation === "number" && Number.isInteger(raw.generation)
        ? raw.generation
        : 0,
  };
}

function sameSourceIdentity(
  left: SemanticOutlineSourceIdentity,
  right: SemanticOutlineSourceIdentity,
): boolean {
  return left.base_id === right.base_id && left.generation === right.generation;
}

export function semanticOutlineValidationContextFromUnknown(
  value: unknown,
): SemanticOutlineValidationContext {
  const raw = mapping(value);
  const units: SemanticOutlineValidationContext["units"] = [];
  for (const item of Array.isArray(raw.units) ? raw.units : []) {
    const unit = mapping(item);
    const unit_id = optionalString(unit.unit_id);
    if (
      unit_id !== null &&
      typeof unit.order_index === "number" &&
      Number.isInteger(unit.order_index) &&
      unit.order_index >= 1
    ) {
      units.push({ unit_id, order_index: unit.order_index });
    }
  }
  const anchors: SemanticOutlineValidationContext["anchors"] = [];
  for (const item of Array.isArray(raw.anchors) ? raw.anchors : []) {
    const anchor = mapping(item);
    const anchor_segment_id = optionalString(anchor.anchor_segment_id);
    const unit_id = optionalString(anchor.unit_id);
    if (anchor_segment_id !== null && unit_id !== null) {
      anchors.push({ anchor_segment_id, unit_id });
    }
  }
  return { source_identity: sourceIdentity(raw.source_identity), units, anchors };
}

export function semanticOutlineValidationInputFromUnknown(
  value: unknown,
): SemanticOutlineValidationInput {
  const raw = mapping(value);
  const attempted_nodes: RawSemanticOutlineNode[] = [];
  for (const item of Array.isArray(raw.attempted_nodes) ? raw.attempted_nodes : []) {
    const node = mapping(item);
    attempted_nodes.push({
      node_id: optionalString(node.node_id),
      parent_node_id: optionalString(node.parent_node_id),
      depth: node.depth,
      title: node.title,
      start_unit_id: optionalString(node.start_unit_id),
      end_unit_id: optionalString(node.end_unit_id),
      start_anchor_segment_id: optionalString(node.start_anchor_segment_id),
      end_anchor_segment_id: optionalString(node.end_anchor_segment_id),
    });
  }
  return {
    field_present: raw.field_present === true,
    requested: raw.requested === true,
    in_flight: raw.in_flight === true,
    worker_failure: raw.worker_failure === true,
    projection_source_identity: sourceIdentity(raw.projection_source_identity),
    attempted_nodes,
  };
}

export function validateSemanticOutlineProjection(
  context: SemanticOutlineValidationContext,
  input: SemanticOutlineValidationInput,
): SemanticOutlineValidationResult {
  const terminal = (
    status: SemanticOutlineStatus,
    reason?: string,
  ): SemanticOutlineValidationResult => ({
    status,
    nodes: [],
    diagnostics: {
      drops: reason === undefined ? [] : [{ node_id: null, reason_code: reason }],
      skipped_node_count: 0,
    },
  });
  if (input.worker_failure) return terminal("failed", "worker_failure");
  if (!input.field_present || (!input.requested && input.attempted_nodes.length === 0)) {
    return terminal("unavailable");
  }
  if (input.in_flight && input.attempted_nodes.length === 0) return terminal("pending");
  if (!sameSourceIdentity(input.projection_source_identity, context.source_identity)) {
    return terminal("stale", "source_mismatch");
  }
  if (input.attempted_nodes.length === 0) return terminal("failed", "empty_attempt");

  const unitOrder = new Map(context.units.map((unit) => [unit.unit_id, unit.order_index]));
  const anchorUnits = new Map(
    context.anchors.map((anchor) => [anchor.anchor_segment_id, anchor.unit_id]),
  );
  const accepted: ValidatedSemanticOutlineNode[] = [];
  const acceptedById = new Map<string, ValidatedSemanticOutlineNode>();
  const droppedIds = new Set<string>();
  const seenIds = new Set<string>();
  const drops: SemanticOutlineDrop[] = [];
  const drop = (
    node: RawSemanticOutlineNode,
    reason_code: string,
    markDropped = true,
  ) => {
    drops.push({ node_id: node.node_id, reason_code });
    if (markDropped && node.node_id !== null) droppedIds.add(node.node_id);
  };

  for (const raw of input.attempted_nodes) {
    if (raw.node_id === null) {
      drop(raw, "missing_node_id");
      continue;
    }
    if (seenIds.has(raw.node_id)) {
      drop(raw, "duplicate_node_id", false);
      continue;
    }
    seenIds.add(raw.node_id);
    const parent = raw.parent_node_id;
    let parentNode: ValidatedSemanticOutlineNode | undefined;
    if (parent !== null) {
      if (droppedIds.has(parent)) {
        drop(raw, "parent_dropped");
        continue;
      }
      parentNode = acceptedById.get(parent);
      if (parentNode === undefined) {
        drop(raw, "invalid_parent");
        continue;
      }
      if (raw.depth !== parentNode.depth + 1) {
        drop(raw, "depth_parent_mismatch");
        continue;
      }
    }
    if (
      typeof raw.depth !== "number" ||
      !Number.isInteger(raw.depth) ||
      raw.depth < 1 ||
      raw.depth > MAX_DEPTH
    ) {
      drop(raw, "depth_out_of_range");
      continue;
    }
    if (parent === null && raw.depth !== 1) {
      drop(raw, "invalid_root_depth");
      continue;
    }
    if (typeof raw.title !== "string") {
      drop(raw, "empty_title");
      continue;
    }
    const title = raw.title.trim();
    if (title.length === 0) {
      drop(raw, "empty_title");
      continue;
    }
    if ([...title].length > TITLE_MAX_CODE_POINTS) {
      drop(raw, "title_too_long");
      continue;
    }
    if (
      raw.start_unit_id === null ||
      raw.end_unit_id === null ||
      !unitOrder.has(raw.start_unit_id) ||
      !unitOrder.has(raw.end_unit_id)
    ) {
      drop(raw, "missing_unit");
      continue;
    }
    const startOrder = unitOrder.get(raw.start_unit_id)!;
    const endOrder = unitOrder.get(raw.end_unit_id)!;
    if (startOrder > endOrder) {
      drop(raw, "inverted_range");
      continue;
    }
    if (parentNode !== undefined) {
      const parentStart = unitOrder.get(parentNode.start_unit_id)!;
      const parentEnd = unitOrder.get(parentNode.end_unit_id)!;
      if (startOrder < parentStart || endOrder > parentEnd) {
        drop(raw, "range_not_nested");
        continue;
      }
    }
    const overlapsSibling = accepted.some(
      (node) =>
        node.parent_node_id === parent &&
        !(
          endOrder < unitOrder.get(node.start_unit_id)! ||
          startOrder > unitOrder.get(node.end_unit_id)!
        ),
    );
    if (overlapsSibling) {
      drop(raw, "range_overlap");
      continue;
    }
    let startAnchor = raw.start_anchor_segment_id;
    let endAnchor = raw.end_anchor_segment_id;
    let invalidAnchor = false;
    if (startAnchor !== null && anchorUnits.get(startAnchor) !== raw.start_unit_id) {
      startAnchor = null;
      invalidAnchor = true;
    }
    if (endAnchor !== null && anchorUnits.get(endAnchor) !== raw.end_unit_id) {
      endAnchor = null;
      invalidAnchor = true;
    }
    if (invalidAnchor) drops.push({ node_id: raw.node_id, reason_code: "invalid_anchor" });
    const node: ValidatedSemanticOutlineNode = {
      node_id: raw.node_id,
      parent_node_id: parent,
      depth: raw.depth,
      title,
      start_unit_id: raw.start_unit_id,
      end_unit_id: raw.end_unit_id,
      start_anchor_segment_id: startAnchor,
      end_anchor_segment_id: endAnchor,
      order_index: accepted.length + 1,
    };
    accepted.push(node);
    acceptedById.set(node.node_id, node);
  }

  const attempted = input.attempted_nodes.length;
  const valid = accepted.length;
  const status: SemanticOutlineStatus =
    attempted === 0 || valid === 0 ? "failed" : valid === attempted ? "ready" : "partial";
  return {
    status,
    nodes: accepted,
    diagnostics: { drops, skipped_node_count: attempted - valid },
  };
}

// ---------------------------------------------------------------------------
// Snapshot wire consumer helpers (null | undefined | object)
// ---------------------------------------------------------------------------

/**
 * Trusted product projection only when a non-null object with ready|partial.
 * `undefined` (legacy omit) and `null` (new backend) are equivalent: no outline.
 * Other status objects are defensively ignored (must not throw or drive UI).
 */
export function hasTrustedSemanticOutline(
  value: ReaderSemanticOutlineProjectionDto | null | undefined,
): value is ReaderSemanticOutlineProjectionDto {
  if (value === null || value === undefined) {
    return false;
  }
  if (typeof value !== "object") {
    return false;
  }
  return value.status === "ready" || value.status === "partial";
}

/**
 * Source-identity key for any future L2 outline cache (same formula as L0/L1).
 * Pure helper only — it does not implement L2 state.
 */
export function buildOutlineSourceIdentityKey(
  baseId: string,
  generation: number,
): string {
  return `${baseId}:${generation}`;
}
